github/resource_github_repository_ruleset_utils.go

```diff
@@ -0,0 +1,256 @@
+	// 		toggleableRule := &github.RepositoryRule{
+	// 			Type:       "update",
+	// 			Parameters: &rawParams,
+	// 		}
+	// 		rulesetRules = append(rulesetRules, toggleableRule)
+	// 	}
+	// }
+
+	if deploymentEnvs, ok := d.GetOk("rule_required_deployments"); ok {
+		requiredDeploymentEnvs := []string{}
+		for _, env := range deploymentEnvs.(*schema.Set).List() {
+			requiredDeploymentEnvs = append(requiredDeploymentEnvs, env.(string))
+		}
+
+		mRequiredDeploymentParams := github.RequiredDeploymentEnvironmentsRuleParameters{
+			RequiredDeploymentEnvironments: requiredDeploymentEnvs,
+		}
+
+		bytes, _ := json.Marshal(mRequiredDeploymentParams)
+		rawParams := json.RawMessage(bytes)
+		requiredDeploymentRule := &github.RepositoryRule{
+			Type:       "required_deployments", // Drop the "rule_", that is only to make the provider implementation easier
+			Parameters: &rawParams,
+		}
+		rulesetRules = append(rulesetRules, requiredDeploymentRule)
+	}
+
+	if v, ok := d.GetOk("rule_pull_request"); ok {
+		vL := v.([]interface{})
+		if len(vL) > 1 {
+			return nil, errors.New("cannot specify rule_pull_request more than one time")
+		}
+		for _, v := range vL {
+			if v == nil {
+				break
+			}
+			m := v.(map[string]interface{})
+			
+			pullRequestRuleParams := github.PullRequestRuleParameters{
+				DismissStaleReviewsOnPush:      m["dismiss_stale_reviews_on_push"].(bool),
+				RequireCodeOwnerReview:         m["require_code_owner_review"].(bool),
+				RequireLastPushApproval:        m["require_last_push_approval"].(bool),
+				RequiredApprovingReviewCount:   m["required_approving_review_count"].(int),
+				RequiredReviewThreadResolution: m["required_review_thread_resolution"].(bool),
+			}
+			bytes, _ := json.Marshal(pullRequestRuleParams)
+			rawParams := json.RawMessage(bytes)
+			requiredDeploymentRule := &github.RepositoryRule{
+				Type:       "pull_request",
+				Parameters: &rawParams,
+			}
+			rulesetRules = append(rulesetRules, requiredDeploymentRule)
+		}
+	}
+
+	if v, ok := d.GetOk("rule_required_status_checks"); ok {
+		vL := v.([]interface{})
+		if len(vL) > 1 {
+			return nil, errors.New("cannot specify rule_required_status_checks more than one time")
+		}
+		for _, v := range vL {
+			if v == nil {
+				break
+			}
+			m := v.(map[string]interface{})
+
+			requiredStatusChecks := m["required_status_checks"].(*schema.Set).List()
+			requiredStatusChecksList := make([]github.RuleRequiredStatusChecks, 0)
+			for _, statusCheck := range requiredStatusChecks {
+
+				// Expect a string of "context:integration_id", allowing for the absence of "integration_id"
+				parts := strings.SplitN(statusCheck.(string), ":", 2)
+				var cContext, cIntegrationId string
+				switch len(parts) {
+					case 1:
+						cContext, cIntegrationId = parts[0], ""
+					case 2:
+						cContext, cIntegrationId = parts[0], parts[1]
+					default:
+						// TODO: What is the prefered way of throwing errors? fmt.Errorf() or errors.New()?
```

Comment:
I have seen both ``fmt.Errorf()`` and `errors.New` being used to throw errors across this project. What is the recommended way?